plugins/experimental/acme/acme.c

```diff
@@ -0,0 +1,347 @@
+acme_process_write(TSCont contp, TSEvent event, AcmeState *my_state)
+{
+  if (event == TS_EVENT_VCONN_WRITE_READY) {
+    char buf[64]; /* Plenty of space for CL: header */
+    int len;
+
+    len = snprintf(buf, sizeof(buf) - 1, "Content-Length: %zd\r\n\r\n", my_state->stat_buf.st_size);
+    my_state->output_bytes += add_data_to_resp(buf, len, my_state);
+    my_state->output_bytes += add_file_to_resp(my_state);
+
+    TSVIONBytesSet(my_state->write_vio, my_state->output_bytes);
+    TSVIOReenable(my_state->write_vio);
+  } else if (TS_EVENT_VCONN_WRITE_COMPLETE) {
+    cleanup(contp, my_state);
+  } else if (event == TS_EVENT_ERROR) {
+    TSError("[%s] acme_process_write: Received TS_EVENT_ERROR", PLUGIN_NAME);
+  } else {
+    TSReleaseAssert(!"Unexpected Event");
+  }
+}
+
+/* Process the accept event from the SM */
+static void
+acme_process_accept(TSCont contp, AcmeState *my_state)
+{
+  my_state->req_buffer = TSIOBufferCreate();
+  my_state->resp_buffer = TSIOBufferCreate();
+  my_state->resp_reader = TSIOBufferReaderAlloc(my_state->resp_buffer);
+  my_state->read_vio = TSVConnRead(my_state->net_vc, contp, my_state->req_buffer, INT64_MAX);
+}
+
+/* Imlement the server intercept */
+static int
+acme_intercept(TSCont contp, TSEvent event, void *edata)
+{
+  AcmeState *my_state = TSContDataGet(contp);
+
+  if (event == TS_EVENT_NET_ACCEPT) {
+    my_state->net_vc = (TSVConn)edata;
+    acme_process_accept(contp, my_state);
+  } else if (edata == my_state->read_vio) { /* All read events */
+    acme_process_read(contp, event, my_state);
+  } else if (edata == my_state->write_vio) { /* All write events */
+    acme_process_write(contp, event, my_state);
+  } else {
+    TSReleaseAssert(!"Unexpected Event");
+  }
+
+  return 0;
+}
+
+/* Read-request header continuation, used to kick off the server intercept if necessary */
+static int
+acme_hook(TSCont contp ATS_UNUSED, TSEvent event ATS_UNUSED, void *edata)
+{
+  TSMBuffer reqp;
+  TSMLoc hdr_loc = NULL, url_loc = NULL;
+  TSCont icontp;
+  AcmeState *my_state;
+  TSHttpTxn txnp = (TSHttpTxn)edata;
+
+  TSDebug(PLUGIN_NAME, "kicking off ACME hook");
+
+  if ((TS_SUCCESS == TSHttpTxnClientReqGet(txnp, &reqp, &hdr_loc)) && (TS_SUCCESS == TSHttpHdrUrlGet(reqp, hdr_loc, &url_loc))) {
+    int path_len = 0;
+    const char *path = TSUrlPathGet(reqp, url_loc, &path_len);
+
+    /* Short circuit the / path, common case */
+    if (!path || path_len < (strlen(ACME_WK_PATH) + 2) || *path != '.' || memcmp(path, ACME_WK_PATH, strlen(ACME_WK_PATH))) {
+      TSDebug(PLUGIN_NAME, "skipping URL path = %.*s", path_len, path);
+      goto cleanup;
+    }
+
+    TSSkipRemappingSet(txnp, 1); /* not strictly necessary, but speed is everything these days */
+
+    /* This request is for us -- register our intercept */
+    icontp = TSContCreate(acme_intercept, TSMutexCreate());
+
+    my_state = make_acme_state();
+    open_acme_file(my_state, path + strlen(ACME_WK_PATH), path_len - strlen(ACME_WK_PATH));
```

Comment:
Where is this file going to be created? ``/var/run`` might be a reasonable choice, or possibly ``TS*DirGet()`` though I don't see an obvious candidate there.